### PR TITLE
Refactor client response-to-error to API helper and add Nessie tests for direct client upload

### PR DESCRIPTION
**Pull request by arielshaqed**
It should be The One True Way To Do It.

### CODE REVIEW

File: pkg/api/helpers/upload.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+)
+
+// ClientUpload uploads contents as a file using client-side ("direct") access to underlying
+// storage.  It requires credentials both to lakeFS and to underlying storage, but
+// considerably reduces the load on the lakeFS server.
+func ClientUpload(ctx context.Context, client api.ClientWithResponsesInterface, repoID, branchID, filePath string, metadata map[string]string, contents io.ReadSeeker) (*api.ObjectStats, error) {
+	resp, err := client.GetPhysicalAddressWithResponse(ctx, repoID, branchID, &api.GetPhysicalAddressParams{
+		Path: filePath,
+	})
+	if err != nil {
```
Comment (itaiad200) [inline pkg/api/helpers/upload.go:25]: Not part of this PR, but is this the recommended way to check for errors from the client?
3 Different checks just to make sure nothing failed..

File: pkg/api/helpers/upload.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+				// BUG(ariels): Unavailable on S3, remove this field entirely
+				//     OR add it to the server staging manager API.
+				Mtime:           time.Now().Unix(),
+				Path:            filePath,
+				PathType:        "object",
+				PhysicalAddress: physicalAddress,
+				SizeBytes:       &size,
+			}, nil
+		}
+		if resp.JSON409 == nil {
```
Comment (itaiad200) [inline pkg/api/helpers/upload.go:102]: This is confusing, but maybe it's just me.
First if checks for the status code, second if checks for the value that comes with the specific status code, without checking the status code first?

File: nessie/system_test.go [hunk 3]
```diff
@@ -75,22 +89,35 @@ func createRepository(ctx context.Context, t *testing.T, name string, repoStorag
...
-		return "", "", err
-	}
-	if resp.StatusCode() != http.StatusCreated {
-		return "", "", fmt.Errorf("%w: %s (%d)", errUploadFailed, resp.Status(), resp.StatusCode())
+	if direct {
+		stats, err := helpers.ClientUpload(ctx, client, repo, branch, objPath, nil, strings.NewReader(objContent))
+		if errors.Is(err, helpers.ErrUnsupportedProtocol) {
+			t.Logf("direct upload not supported yet (%s), switch to use gateway", err)
+			directUploadSupported = false
+			return uploadFileRandomDataAndReport(ctx, t, repo, branch, objPath, false)
```
Comment (itaiad200) [inline nessie/system_test.go:105]: This is an indirect way to understand if the current settings support direct upload or not.
Since we're trying to test the actual system, it might ignore real errors. For instance, s3 returns `ErrUnsupportedProtocol`. 

Any chance we can pass it as a test argument and expects all actions to succeed (or fail) according to the prior expectation, and not one inferred from the first call?

File: nessie/system_test.go [hunk 4]
```diff
@@ -110,8 +137,8 @@ func uploadContent(ctx context.Context, repo string, branch string, objPath stri
 	}, w.FormDataContentType(), &b)
 }
 
-func uploadFileRandomData(ctx context.Context, t *testing.T, repo, branch, objPath string) (checksum, content string) {
-	checksum, content, err := uploadFileRandomDataAndReport(ctx, repo, branch, objPath)
+func uploadFileRandomData(ctx context.Context, t *testing.T, repo, branch, objPath string, direct bool) (checksum, content string) {
+	checksum, content, err := uploadFileRandomDataAndReport(ctx, t, repo, branch, objPath, direct && directUploadSupported)
```
Comment (itaiad200) [inline nessie/system_test.go:141]: Maybe check the value of `directUploadSupported` in the inner method instead of maybe missing it occasionally from other callers of `uploadFileRandomDataAndReport`.

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Really nice work with enabling client upload and testing it.

A little bit worried about the reflection part and deciding when to use the direct upload from Nessie.

File: nessie/system_test.go (hunk 4 above)
Comment (arielshaqed) [reply nessie/system_test.go]: This no longer exists, and `direct` is piped right through.

File: pkg/api/helpers/upload.go (hunk 1 above)
Comment (arielshaqed) [reply pkg/api/helpers/upload.go:25]: No idea.  @nopcoder WDYT?

File: pkg/api/helpers/upload.go (hunk 2 above)
Comment (arielshaqed) [reply pkg/api/helpers/upload.go:102]: Yes, AFAIU.  The code gen creates a response type with a separate field for each possible status code.  That field will be `nil` if the status code is different.  The converse need not be true, so above we _do_ test the status code because we ignore all response fields.
@nopcoder to confirm.

File: pkg/api/helpers/errors.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+// ResponseAsError returns a ErrRequestFailed wrapping a response from the server.  It
+// searches for a non-nil unsuccessful HTTPResponse field and uses its message, along with a
+// Body that it assumes is an api.Error.
+func ResponseAsError(response interface{}) error {
+	r := reflect.Indirect(reflect.ValueOf(response))
+	if !r.IsValid() {
+		return fmt.Errorf("%w: bad type: cannot indirect", ErrRequestFailed)
+	}
+
+	f := r.FieldByName("HTTPResponse")
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply pkg/api/helpers/errors.go:39]: Done, sort-of.  The docs mean that _`r`_ has to have a struct type.  It does, so that's OK.

Modified anyway to test for the kind of `r` -- it will be an internal (logic) error reported as an error, but probably less scary than a `panic`.

File: pkg/api/helpers/errors.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	f = r.FieldByName("Body")
+	if f.IsValid() && f.Type().Kind() == reflect.Slice && f.Type().Elem().Kind() == reflect.Uint8 {
+		body := f.Bytes()
+		var apiError api.Error
+		if json.Unmarshal(body, &apiError) == nil && apiError.Message != "" {
+			message = apiError.Message
+		}
+	}
+
+	return fmt.Errorf("%w: %s", ErrRequestFailed, message)
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply pkg/api/helpers/errors.go:65]: Done

[the remark this answers is not in the document]
Comment (arielshaqed) [reply pkg/api/helpers/errors.go:65]: No.  I want to get "request failed: invalid storage type s3: current block adapter is gcp".  With the above change I get "invalid storage type s3: current block adapter is gcp: request failed" which is not as nice.

[the remark this answers is not in the document]
Comment (arielshaqed) [reply pkg/api/helpers/errors.go:65]: Including it.

File: pkg/api/helpers/errors_test.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func TestResponseAsError(t *testing.T) {
+	cases := []struct {
+		name     string
+		response interface{}
+		message  string // non-empty to match return error; empty to signal no error
+	}{
+		{"no HTTPResponse field", &struct{ A int }{17}, "no HTTPResponse"},
+		{"OK", &Response{&http.Response{StatusCode: 234}}, ""},
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply pkg/api/helpers/errors_test.go:27]: Discussed f2f; `reflect.Value.Field` has unexpected behaviour... and this is OK.

**Review by arielshaqed — COMMENTED** [review]
Thanks!

PTAL...

**Review by itaiad200 — APPROVED** [review]

**Comment by arielshaqed** [comment]
Thanks!  Pulling once tests pass...
